### ISSUE TITLE
Transmutation Table forgetting items.

### ISSUE THREAD

**Issue by Oraiz** [sinkillerj/ProjectE#1212]
Latest 1.7.10 version with Forge .1614, seems that after closing and restarting client, a lot of items (so far it seems only vanilla) are forgotten. I have plenty of available EMC.

**Comment by sinkillerj** [comment]
Logs please.

**Comment by MaPePeR** [comment]
This has been reported multiple times, but we were not able to reproduce it or find the cause. I don't think there will be log messages. And even if there are: the bug already happened. You would need messages from the exact time the items disappeared.

**Comment by Oraiz** [comment]
Yeah I've searched to try and find a solution, but there seem to only be people having the same issue without resolution. I've no idea what's causing it, I'm not sure which logs you would need? Let me know and I'll upload any of relevance. :)

**Comment by sinkillerj** [comment]
If you were on a server we need your server logs, otherwise the logs for the client. Both store multiple logs in its mods folder, if you open them up it should be fairly easy to find the times you played before and after the items disappeared as they are time and date stamped.

**Comment by williewillus** [comment]
These reports all started coming in at some defined point in time meaning that something screwed up in the last few releases. I might take a look in 1.9. I'm not going to touch the 1.7 codebase again :P

Also, there won't be anything in logs I don't think

**Comment by williewillus** [comment]
Though, in 1.9 that whole system is changing again, so it might even be gone

**Comment by Oraiz** [comment]
Hey, thanks for the replies. I gave myself the tome to learn everything, nothing seems to have disappeared since. I've just been not using anything I haven't learned before cheating in.

Are you still actively working on the 1.7.10 version? Would trying to solve this be helpful for you? If not, you don't need to waste your time just for me. :)

**Comment by williewillus** [comment]
I'm not, I'm solely focused on 1.9 and 1.8 (mostly 1.9), the other devs are the ones that work on 1.7

**Comment by sinkillerj** [comment]
Right now I'm working on 1.7, so logs are certainly helpful. Either way if we find the cause in 1.7 it could help if the code has the same problem in 1.8/1.9.

**Comment by MaPePeR** [comment]
Useful information might be: single player or multiplayer? Did the items disappear while playing or did you only notice them missing after a restart?

**Comment by Oraiz** [comment]
[2016-04-05-6.log.gz](https://github.com/sinkillerj/ProjectE/files/209967/2016-04-05-6.log.gz)
[2016-04-05-2.log.gz](https://github.com/sinkillerj/ProjectE/files/209966/2016-04-05-2.log.gz)
[2016-04-05-3.log.gz](https://github.com/sinkillerj/ProjectE/files/209968/2016-04-05-3.log.gz)
[2016-04-05-4.log.gz](https://github.com/sinkillerj/ProjectE/files/209969/2016-04-05-4.log.gz)
[2016-04-05-5.log.gz](https://github.com/sinkillerj/ProjectE/files/209970/2016-04-05-5.log.gz)

These are the logs from the first day, before I cheated in the tome, I think. I can upload more if you need them? Also, if these are the wrong ones please let me know and I'll upload more.

Single player, added to FTB Infinity Evolved: Skyblock, along with a few other mods that I 'can't play without'. 
Been playing my own custom pack before with no problem at all - I would guess it may be something to do with all the MineTweaking? But then it's an uneducated guess. :p

I'm sorry, I can't remember if they were disappearing while I was playing; I'm fairly certain it was after a client restart though.

**Comment by sinkillerj** [comment]
Do you save and quit or simply close your client? In some cases closing the client in single player can cause data loss.

**Comment by Oraiz** [comment]
Yes, save and close. It has started happening again after a few days of playing with the tome cheated in too.

**Comment by williewillus** [comment]
it's probably the tome then

**Comment by MaPePeR** [comment]
I don't think its the tome itself. I think its triggered by a lot of learned items (What then is triggered by learning the tome)

**Comment by Oraiz** [comment]
It isn't the tome, as far as I know anyway, because it was happening before I used it (that's why I tried with it too). 

I don't think it would be item limits, as before I only had the majority of the 'easy to get' vanilla items, and maybe 20-30 modded items, when I noticed saplings not being there when I searched for one. :)

**Comment by Oraiz** [comment]
I've since added PlayerAPI to my custom pack, since the portal to The Last Millennium wasn't working, and I heard that this solved it. When I accessed my table, a huge amount of items were forgotten, so I think this may have something to do with it?

**Comment by MaPePeR** [comment]
This might be caused by an item ID change. 

Do we save item ids instead of unique names for knowledge?! 

No other report mentioned that the knowledge loss is linked to adding/removing Mods.

**Comment by williewillus** [comment]
Itemstacks are serialized the way they always are in normal usage for knowledge

**Comment by MaPePeR** [comment]
Maybe we need to change the serialization for persistent storage to use the unique id instead of the numerical item ID?

**Comment by destrofer** [comment]
Well, today I noticed the issue and investigated it a bit. I believe no logs will be helpful here. Instead I will tell my experience and steps which allowed me to reproduce the issue:

1. Have 2 items without known EMC value (for example item A and item B), where A is craftable only from items with known EMC, and B is craftable from A and optionally any other items with known EMC.
2. At this point when you start the game A will have EMC calculated, but B won't.
3. Executing command "/projecte reloadEMC" calculates the EMC for item B since A already had it calculated before that.
4. Add item B to table.
5. Quit the game. Just doing "save and quit" and then reloading the world will have EMC for item B and it will not disappear from table, but if you exit the game itself and restart it, everything will be back to step 1 with item B being forgotten.

As a coder myself I might give you a hint where the problem is. Sorry, I haven't checked the code myself, so I can only give assumptions.

I assume the issue is with the order of calculations in EMC calculation algorithm.

For example if item A was craftable from item C with known EMC, then there are two possible scenarios.

First scenario:
- A is calculated from C as expected
- B is calculated from A
- everyone's happy and all items have EMC

Second scenario
- B is not calculated since A has no EMC yet
- A is calculated from C as expected
- that gives us the headache

The scenario really depends on the order of items in the list.

So, if it really is the case, then I can suggest two solutions:

1. Repeat the whole calculation cycle until it does not change anything
```
do {
    changed = false;
    for( all items ) {
        if( item EMC needs calculation) {
            calculate EMC;
            if( calculated successfully ) {
                assign EMC;
                changed = true;
            }
        }
    }
} while(changed);
```

2. Use dependency/stack approach:
```
calculateEMC(item) {
    for( all items this item is crafted from ) {
        calculateEMC(dependency);
    }
}
```

First solution is simple to implement, but may be slower since it might need to do several passes over all items.

Second solution looks simple in pseudo code, but there lies another issue: crafting recipes may have  looping dependencies. For example A is craftable from B, B is craftable from A. As a result this might require a lot of memory to track what items are already being calculated to prevent going into an infinite loop and crash due to stack overflow.

So that is that. As to what to do with this is for you to decide, but do I hope it will get fixed soon.

P.S. My friend went nuts when he heard about ProjectE. Thanks for bringing us back the old times :)

**Comment by MaPePeR** [comment]
The emc calculation is already repeating until nothing changes.

**Comment by destrofer** [comment]
Well then at least my steps to reproduce might help you a little bit, since it still happens.

Mod version: PE1.9.5
Minecraft 1.7.10
Forge 1614

In my case actual items are from portal gun mod:
Item A: Miniature Black Hole
item B: Portal Gun

Portal gun is getting forgotten every time I restart the game (not the world).

**Comment by sinkillerj** [comment]
Portal has had issues in the past and may still #897

**Comment by destrofer** [comment]
I'm playing around with ProjectE and portal gun right now and noticed one behaviour: once you shoot the orange portal, the selected portal gun is replaced with the orange one that has different ID. Since it is impossible to craft orange portal gun, it has no EMC even after `/projecte reloadEMC`. The default one still has same issues as I described earlier. If portal gun mod registers its "gun" items in the later mod loading step than ProjectE, then nothing can be done except informing them of this issue. It is just a bit strange, since other items of the same mod get their EMC calculated and it is hard to believe they register them in a separate mod loading step.

Furthermore I am a bit unsure about that, but yesterday I noticed "forgetting" one of items from other mod. The problem is that at the time I thought that I may have forgotten to add it to the table. Now I won't be able to remember what that item it was and confirm it for sure. However if it was also similar case to protal gun mod, then running around and asking all mod makers to register their items on specific steps would be impossible.

**Comment by MaPePeR** [comment]
Portal Gun does register its Crafting Recipe at a very late step in the loading of the game state, because the crafting recipes depend on Game Difficulty. 

To fix the issues with portal gun:

* Start the game in the difficulty you want the recipes to be in.
* Open the mapper.cfg and enable pregenerated EMC values
* type the `/projecte reloademc` command in the game that is still running
* Now EMC values will never ever change again. Even if you add new mods you need to repeat these steps to make them get EMC values

There is nothing we can do about this, but pregenerated EMC is there to fix this and other problems.

I don't think the knowledge loss is related to EMC generation.

**Comment by Bersham** [comment]
Okay, here it goes again...This started happening to me just as I logged today...
I've been playing this world for a few days now without a problem. I play with cheats on to add custom EMC values to items every now and then.

Anyway, I have over 5 million EMC in the Transmutation Table. When I logged in today I just walked around and decided to start building shit. So I went into the Transmutation Table and searched for planks (Everything seemed to be in there. I had hundreds of things saved in there). I Shift+Right-Clicked the planks and assumed that the three sets of different planks went into my inventory. But then I exited out of the Transmutation Table and looked into my inventory and didn't see the bloody planks. I thought I miss-clicked, but when I checked the Transmutation Table, most of the items were gone. I had 3 pages of stuff from Minecraft and a couple of mods still inside the Transmutation Table. The EMC inside the Transmutation Table stayed the same.

As far as I remember, yesterday I set custom EMC for Redstone Energy Cell Frame (Full) and Resonant Energy Cell Frame (Full) from Thermal Expansion. I had to cheat in a few Dark Matter pickaxes because Glenn's Gases' Lanterns will except any item when Right-Clicking on them. The Lanterns deleted my pickaxes a few times.

I never updated the mods while playing this world, so item named and stuff couldn't have changed right?. This is as fresh a world I could make.

Minecraft 1.7.10
Forge 10.13.4.1614
ProjectE 1.10




[Latest-June 1](https://pastenib.com/v/zQSWEBfDbJ42a9amjBcx)
[Latest-May 31 #2](https://pastenib.com/v/t7TfihcPxeGwylp1Didj)
[Latest-May 31 #1](https://pastenib.com/v/vR3X0iJN0ZIAYrz9OcEU)

**Comment by sinkillerj** [comment]
Should be fixed when 1.10 is released.

**Comment by hron84** [comment]
@sinkillerj this issue happens me again at 1.10 with SkyExchange modpack. Previously, I had no issues when I went to the nether and back, but now transmutation table started to forget stuffs. Sometimes it goes back, sometimes not.

**Comment by williewillus** [comment]
Fixed in next release for 1.10

**Comment by hron84** [comment]
@williewillus Thanks for the quick reply. Is there any workaround for it?

**Comment by williewillus** [comment]
relogging should work, and make sure you're on the latest 1.10 version

**Comment by hron84** [comment]
@williewillus thanks, i'll check it. I'm on the latest, i mean curse client does not display an upgrade sign next to your mod.

**Comment by jenthura** [comment]
Just adding my two cents. I'm on the latest version of MC. Every time I enter or leave the Nether my transmutation tablet forgets all learned items and loses all stored EMC. I've fixed it by going into creative and grabbing about 20-30 items I learned before and adding them in. It seems to be random, but when I get to around 20 items it suddenly just remembers all the recipes again. Very bizarre.

[2016-06-09-1.log.gz](https://github.com/sinkillerj/ProjectE/files/659627/2016-06-09-1.log.gz)

**Comment by williewillus** [comment]
latest as in the one release today/yesterday?

**Comment by hron84** [comment]
@jenthura before starting re-learning stuffz, please do a logout/login cylce instead. In most cases it fixes the problem. However, relearned stuffz get lost also the grabbed EMC goes to waste when you use trans talbe in that state.

**Comment by jenthura** [comment]
@williewillus MC 1.10.2 and PE1.1.0B

**Comment by sinkillerj** [comment]
@jenthura Time to upgrade then, you are behind a version.

**Comment by Prometheanrazer** [comment]
Will this get fixed for 1.7.10 or is that out of negotiation?
a year later the error is still a thing on 1.7.10

for me seemed to happen after adding Aether II so probably item ID's changing

**Comment by sinkillerj** [comment]
1.7 has been unsupported for a very long time, no further updates are in the works for it.
